src/ophys_etl/transforms/sine_dewarp.py

```diff
@@ -0,0 +1,534 @@
+    aL = float(aLstr)
+
+    aRstr = data.get("aR", None)
+    if aRstr is None:
+        raise KeyError(f"parameter aR not determined for {equipment_name}")
+    aR = float(aRstr)
+
+    bLstr = data.get("bL", None)
+    if bLstr is None:
+        raise KeyError(f"parameter bL not determined for {equipment_name}")
+    bL = float(bLstr)
+
+    bRstr = data.get("bR", None)
+    if bRstr is None:
+        raise KeyError(f"parameter bR not determined for {equipment_name}")
+    bR = float(bRstr)
+
+    logging.debug(str(equipment_name))
+    logging.debug(f"aL: {aL}   bL: {bL}")
+    logging.debug(f"aR: {aR}   bR: {bR}")
+
+    return input_h5, output_h5, aL, aR, bL, bR
+
+
+def make_output_file(output_file, output_dataset,
+                     FOVwidth, movie_shape, movie_dtype):
+    """
+
+    """
+    # Remove old output if it exists and create new output file
+    if os.path.exists(output_file):
+        os.remove(output_file)
+
+    dewarped_file = h5py.File(output_file, 'w')
+    if FOVwidth == 512:
+        dewarped_file.create_dataset(
+            output_dataset, shape=movie_shape, dtype=movie_dtype
+        )
+    else:
+        out_shape = [movie_shape[0], movie_shape[1], 512 - R - L]
+        dewarped_file.create_dataset(
+            output_dataset, shape=out_shape, dtype=movie_dtype
+        )
+    dewarped_file.close()
+
+
+def run_dewarping(FOVwidth, noise_reduction, threads,
+                  input_file, input_dataset, output_file, output_dataset):
+    """
+    Gets information about the movie and the specified dewarping method,
+    then uses multiprocessing to dewarp each frame of the movie at once, while
+    tracking the progress.
+
+    Parameters
+    ----------
+    FOVwidth: int
+        Field of View width, a parameter of the experiment.
+    Returns
+    -------
+    """
+
+    # Get statistics about the movie
+    input_h5_file = h5py.File(input_file, 'r')
+    movie = input_h5_file[input_dataset]
+
+    movie_shape = movie.shape
+    movie_dtype = movie.dtype
+    T, y_orig, x_orig = movie.shape
+
+    # IMPORTANT!: estmated modevalue is from 8bit img but convert back to 16bit
+    xtable = create_xtable(movie, aL, aR, bL, bR, noise_reduction)
+
+    make_output_file(output_file, output_dataset,
+                     FOVwidth, movie_shape, movie_dtype)
+    input_h5_file.close()
+
+    start_time = time.time()
+    with multiprocessing.Pool(threads) as pool, h5py.File(output_h5, "a") as f:
+        fn = functools.partial(
+            xdewarp,
+            input_file=input_file,
+            input_dataset=input_dataset,
+            FOVwidth=FOVwidth,
+            xtable=xtable,
+            noise_reduction=noise_reduction
+        )
+
+        for frame, dewarped_frame in pool.imap_unordered(fn,
+                                                         range(T),
+                                                         chunksize=100):
+            f[output_dataset][frame, :, :] = dewarped_frame
```

Comment:
I think reading 1 frame at a time is going to be super inefficient based on the way hdf5 files are organized.

See: https://support.hdfgroup.org/HDF5/doc/Advanced/Chunking/

Instead I would suggest something like:
```
input_h5_file = h5py.File(input_file, 'r')
start_time = time.time()

for start_indx in range(0, T, chunk_size):
    remaining_frames = T - start_indx
    end_indx = start_indx + chunk_size if (remaining_frames > chunk_size) else start_indx + remaining_frames

    movie_chunk = input_h5_file[input_dataset][start_indx:end_indx]

    with mp.Pool(threads) as pool, h5py.File(output_h5, "a") as out_h5:
         # Create a starmap iterable see: https://docs.python.org/3.7/library/multiprocessing.html#multiprocessing.pool.Pool.starmap
         warp_inputs = [(image_frame, FOV_width, xtable, noise_reduction) for image_frame in movie_chunk]
         results = pool.starmap(xdewarp, warp_inputs)
         f[output_dataset][start_indx:end_indx] = results

input_h5_file.close()
end_time = time.time()
logging.debug(f"Elapsed time (s): {end_time - start_time}")

```

Note this is untested code so may not work right out of box.

Reply:
The reading actually didn't seem to be a bottleneck for time. But I do think that chunking up the data manually and using a doing a number of starmaps serially will help us with the current problem of returning too much at once from the imap!